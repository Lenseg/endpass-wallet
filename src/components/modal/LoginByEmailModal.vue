<template>
  <v-modal @close="handleClose">
    <template slot="header">Login/Register</template>

    <v-form @submit="handleSubmit">
      <v-input v-model="email"
               label="Email"
               name="email"
               validator="required|email"
               placeholder="Your email"
               :disabled="isLoading" />

      <v-checkbox v-model="termsAccepted">
        I accept <a href="https://endpass.com/terms/" target="_blank">Terms of Service</a>
        and <a href="https://endpass.com/privacy/" target="_blank">Privacy Policy</a>
      </v-checkbox>

      <v-button className="is-primary is-medium"
                :disabled="!termsAccepted"
                :loading="isLoading">Send</v-button>
    </v-form>
  </v-modal>
</template>

<script>
import VModal from '@/components/ui/VModal';
import VForm from '@/components/ui/form/VForm';
import VInput from '@/components/ui/form/VInput';
import VButton from '@/components/ui/form/VButton';
import VCheckbox from '@/components/ui/form/VCheckbox';

export default {
  name: 'login-by-email-modal',
  props: {
    isLoading: {
      type: Boolean,
      default: false,
    },
  },
  data:() => ({
    email: '',
    termsAccepted: false,
  }),
  methods: {
    handleSubmit() {
      this.$emit('confirm', this.email);
    },
    handleClose() {
      this.$emit('close');
    },
  },
  components: {
    VModal,
    VForm,
    VInput,
    VButton,
    VCheckbox,
  },
};
</script>
