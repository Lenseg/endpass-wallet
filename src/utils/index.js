export const kebabToCamel = input =>
  input.replace(/-([a-z])/g, g => g[1].toUpperCase());
