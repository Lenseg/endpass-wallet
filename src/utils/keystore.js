import keythereum from 'keythereum';
import bs58check from 'bs58check';
import { kdfParams } from '@/config';

// Monkey patch keythereum to skip generating address for private keys
// This allows us to encrypt private keys of arbitrary length, and
// conforms better to the Ethereum keystore V3 spec, which does not include
// the address for privacy reasons.
// See https://github.com/ethereum/wiki/wiki/Web3-Secret-Storage-Definition#alterations-from-version-1
keythereum.privateKeyToAddress = function(pk) {
  return '';
};

export default {
  // Encrypts a private key Buffer into a V3 keystore object
  // The exported keystore does NOT include an address
  encrypt(password, privateKey) {
    // Generate random salt and iv for each encryption
    let dk = keythereum.create();
    let options = {
      kdf: kdfParams.kdf,
      kdfparams: kdfParams,
    };
    let encrypted = keythereum.dump(
      password,
      privateKey,
      dk.salt,
      dk.iv,
      options,
    );
    delete encrypted.address;
    return encrypted;
  },

  // Decrypts a V3 keystore object into a private key Buffer
  decrypt(password, keystore) {
    return keythereum.recover(password, keystore);
  },

  // Encode to Base58Check
  encodeBase58(key) {
    return bs58check.encode(key);
  },

  // Decode from Base58Check
  decodeBase58(key) {
    return bs58check.decode(key);
  },
};
