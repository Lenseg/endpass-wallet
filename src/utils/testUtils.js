const toKebab = string =>
  string.replace(/([a-z])([A-Z])/g, '$1-$2').toLowerCase();

export function generateStubs(Component) {
  return Object.values(Component.components).reduce((stubs, stubComponent) => {
    let dashName;

    if (stubComponent.name) {
      dashName = toKebab(stubComponent.name);
    }

    stubs[dashName] = {
      render(createElement) {
        return createElement(dashName, this.$slots.default);
      },
    };

    return stubs;
  }, {});
}
