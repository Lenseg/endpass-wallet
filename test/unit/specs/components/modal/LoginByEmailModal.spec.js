import { shallow, mount, createLocalVue } from '@vue/test-utils';

import LoginByEmailModal from '@/components/modal/LoginByEmailModal';
import { generateStubs } from '@/utils/testUtils';

describe('LoginByEmailModal', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(LoginByEmailModal, {
      stubs: generateStubs(LoginByEmailModal),
    });
  });

  describe('render', () => {
    it('should be a Vue component', () => {
      expect(wrapper.name()).toBe('login-by-email-modal');
      expect(wrapper.isVueInstance()).toBeTruthy();
    });

    it('should render component', () => {
      expect(wrapper.element).toMatchSnapshot();
    });

    describe('the send button', () => {
      it('should change "disable" attribute', () => {
        wrapper.setData({
          termsAccepted: true,
        });

        expect(wrapper.find('v-button').attributes().disabled).toBeFalsy();

        wrapper.setData({ termsAccepted: false });
        expect(wrapper.find('v-button').attributes().disabled).toBeTruthy();
      });
    });
  });

  describe('props', () => {
    describe('isLoading', () => {
      it('should correctly change button "loading" property', () => {
        wrapper.setProps({
          isLoading: false,
        });
        expect(wrapper.find('v-button').attributes().loading).toBeFalsy();

        wrapper.setProps({
          isLoading: true,
        });
        expect(wrapper.find('v-button').attributes().loading).toBeTruthy();
      });
    });
  });

  describe('methods', () => {
    describe('handleSubmit', () => {
      it('should trigger "confirm" event', () => {
        const email = 'email';

        wrapper.setData({ email });
        wrapper.vm.handleSubmit();

        expect(wrapper.emitted().confirm).toEqual([[email]]);
      });
    });

    describe('handleClose', () => {
      it('should trigger "close" event', () => {
        wrapper.vm.handleClose();

        expect(wrapper.emitted().close).toEqual([[]]);
      });
    });
  });
});
