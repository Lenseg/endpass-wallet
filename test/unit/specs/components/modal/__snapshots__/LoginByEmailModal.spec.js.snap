// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginByEmailModal render should render component 1`] = `
<v-modal>
   
  <v-form>
    <v-input
      label="Email"
      name="email"
      placeholder="Your email"
      validator="required|email"
    />
     
    <v-checkbox>
      
      I accept 
      <a
        href="https://endpass.com/terms/"
        target="_blank"
      >
        Terms of Service
      </a>
      
      and 
      <a
        href="https://endpass.com/privacy/"
        target="_blank"
      >
        Privacy Policy
      </a>
    </v-checkbox>
     
    <v-button
      classname="is-primary is-medium"
      disabled="true"
    >
      Send
    </v-button>
  </v-form>
</v-modal>
`;
